File: src/test/java/ru/javawebinar/topjava/TestMatcher.java
package ru.javawebinar.topjava;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class TestMatcher<T> {

    private String[] fields;

    private TestMatcher(String... fields) {
        this.fields = fields;
    }

    public static <T> TestMatcher<T> usingFieldsComparator(String... fields) {
        return new TestMatcher<>(fields);
    }
    public void assertMatch(T actual, T expected) {
        assertThat(actual).isEqualToIgnoringGivenFields(expected, fields);
    }

    public void assertMatch(Iterable<T> actual, T... expected) {
        assertMatch(actual, List.of(expected));
    }

    public void assertMatch(Iterable<T> actual, Iterable<T> expected) {
        assertThat(actual).usingElementComparatorIgnoringFields(fields).isEqualTo(expected);
    }
}

File: src/test/java/ru/javawebinar/topjava/TestMatcher.kt
package ru.javawebinar.topjava

import org.assertj.core.api.Assertions
import ru.javawebinar.topjava.model.Meal
import ru.javawebinar.topjava.model.User


class TestMatcher<T> private constructor(private vararg val fields: String) {

    companion object {
        fun <T> usingFieldsComparator(vararg fields: String) = TestMatcher<T>(*fields)
    }

    fun assertMatch(actual: T, expected: T) {
        Assertions.assertThat(actual).isEqualToIgnoringGivenFields(expected, *fields)
    }

    fun assertMatch(actual: Iterable<T>, obj1: T, vararg objs: T) {
        assertMatch(actual, listOf(obj1, *objs))
    }

    fun assertMatch(actual: Iterable<T>, expected: Iterable<T>) {
        Assertions.assertThat(actual).usingElementComparatorIgnoringFields(*fields).isEqualTo(expected)
    }

}

val USER_MATCHER = TestMatcher.usingFieldsComparator<User>("registered", "roles", "_meals")

val MEAL_MATCHER = TestMatcher.usingFieldsComparator<Meal>("user")